src/cloud-api-adaptor/pkg/securecomms/wnssh/wnssh.go

```diff
@@ -0,0 +1,310 @@
+	logger.Printf("Using PP SecureComms: InitSshClient version %s", sshutil.PpSecureCommsVersion)
+
+	// Read WN Secret
+	wnPrivateKey, wnPublicKey, err := kubemgr.KubeMgr.ReadSecret(sshutil.ADAPTOR_SSH_SECRET)
+	if err != nil {
+		// auto-create a secret
+		wnPrivateKey, wnPublicKey, err = kubemgr.KubeMgr.CreateSecret(sshutil.ADAPTOR_SSH_SECRET)
+		if err != nil {
+			return nil, fmt.Errorf("failed to auto create WN secret: %w", err)
+		}
+	}
+	if len(wnPrivateKey) == 0 {
+		return nil, fmt.Errorf("missing keys for PeerPod")
+	}
+	// Create the Signer for this private key.
+	signer, err := ssh.ParsePrivateKey(wnPrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse private key: %v", err)
+	}
+
+	kbscPrivateKey, _, err := kubemgr.KubeMgr.ReadSecret(sshutil.KBS_CLIENT_SECRET)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read KBS client secret: %w", err)
+	}
+
+	kc := InitKbsClient(kbsAddress)
+	err = kc.SetPemSecret(kbscPrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("KbsClient - %v", err)
+	}
+
+	wnSecretPath := "default/sshclient/publicKey"
+	logger.Printf("Updating KBS with secret for: %s", wnSecretPath)
+	err = kc.PostResource(wnSecretPath, wnPublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to PostResource WN Secret: %v", err)
+	}
+
+	sshClient := &SshClient{
+		kc:              kc,
+		wnSigner:        &signer,
+		inboundStrings:  inbound_strings,
+		outboundStrings: outbound_strings,
+		sshport:         sshport,
+	}
+
+	return sshClient, nil
+}
+
+func (ci *SshClientInstance) GetPort(name string) string {
+	var ok bool
+	var inPort string
+	inPort, ok = ci.inboundPorts[name]
+	if !ok {
+		return ""
+	}
+	return inPort
+}
+func (ci *SshClientInstance) DisconnectPP(sid string) {
+
+	ci.inbounds.DelAll()
+
+	// Cancel the VM connection
+	ci.cancel()
+	ci.wg.Wait()
+	logger.Print("SshClientInstance DisconnectPP success")
+
+	// Remove peerPod Secret named peerPodId
+	kubemgr.KubeMgr.DeleteSecret(PpSecretName(sid))
+}
+
+func (c *SshClient) InitPP(ctx context.Context, sid string, ipAddr []netip.Addr) *SshClientInstance {
+	// Create peerPod Secret named peerPodId
+	var ppPublicKey, ppPrivateKey []byte
+	var err error
+	var kubernetesPhase bool
+
+	// Try reading first in case we resume an existing PP
+	logger.Printf("InitPP read/create PP secret named: %s", PpSecretName(sid))
+	ppPrivateKey, ppPublicKey, err = kubemgr.KubeMgr.ReadSecret(PpSecretName(sid))
+	if err != nil {
+		ppPrivateKey, ppPublicKey, err = kubemgr.KubeMgr.CreateSecret(PpSecretName(sid))
+		if err != nil {
+			logger.Printf("Failed to create PP secret: %v", err)
+			return nil
+		}
+	} else {
+		// we already have a store secret for this PP
+		kubernetesPhase = true
+	}
+
+	// >>> Update the KBS about the SID's Secret !!! <<<
+	sidSecretPath := fmt.Sprintf("default/pp-%s/privateKey", sid)
+	logger.Printf("Updating KBS with secret for: %s", sidSecretPath)
+	err = c.kc.PostResource(sidSecretPath, ppPrivateKey)
```

Comment:
I think you'll need to do a cleanup using `defer` to handle error cases later on in the code. Otherwise there will be stale pp secrets in the KBS.

Reply:
We will need to add DeleteResource() support in KBS first, we need it also for when a PP is terminated.
Leaving it to a different PR, after the KBS is enhanced with a delete option.

Reply:
Let's document this for now in the SecureComms.md